### ISSUE TITLE
Markdown rendering support in commit messages

### ISSUE THREAD

**Issue by KvanTTT** [gitextensions/gitextensions#6509]
## Feature description

It would be great to add Markdown support (**bold**, *italic*, [link](#), and another syntax) in Git messages.

## Environment

- Git Extensions 3.0.2.5232
- Build 23b6f51905006ccdde8701591df706284b4155dc
- Git 2.19.2.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.3163.0
- DPI 96dpi (no scaling)

**Comment by mast-eu** [comment]
I don't think this is possible. There are dozens of GUI clients on all major OS, several websites (github, gitlab, etc) and even the Linux shells and Windows command line to be considered. All of them can be used to show git commit messages. Thus any flavour of formatted messages would need to be supported by all of these.
Otherwise we end up with having a single app (or, at best, a few apps) supporting markdown-formatted messages, while all others would show the raw message with format tags.

**Comment by KvanTTT** [comment]
You are right, but Markdown can be used as a plain text. It's just an additional feature.

**Comment by gerhardol** [comment]
What is the specific feature request?
To render MD if detected in the commit message?
Commit tab only?
What specific is supported on other sites?

The better you can describe the the steps you want to see and how this plays together with other Git clients (including web), the higher possibility that someone will support (and maybe implement) this.

**Comment by KvanTTT** [comment]
> What is the specific feature request?

Improving UX by providing clickable links and by highlighting **bold**, *italic*, `code lines` in commit messages. If a user uses multiline commit messages, headers, quotes, lists, code fragments, and another Markdown syntax can be used. Sometimes it's would be great to go to some entity definition by link and now it's impossible to do.

> To render MD if detected in the commit message?
> Commit tab only?

Both. It can be an additional option "Render Markdown in commit messages" in Settings. Other Git clients don't support such feature yet as I know.

> What specific is supported on other sites?

GitHub recognizes user mentions (@KvanTTT), links to issues and pull requests (#6509), pointers to commits (23b6f51905006ccdde8701591df706284b4155dc), hyperlinks (https://github.com/gitextensions/gitextensions).

**Comment by gerhardol** [comment]
> GitHub recognizes user mentions (@KvanTTT), links to issues and pull requests (#6509), pointers to commits (23b6f51), hyperlinks (https://github.com/gitextensions/gitextensions).

Not for users (as GE does not have the GH user database), but email and the others:
https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#git-extensions-revision-links

To support this, the WinForms RichTextBox that GE uses for CommitInfo will need to parse the text. This must be easy to disable. The commit message aalso is displayed in tooltips etc where html support is not currently available.

Whatever is done, the MD support will be a little hacky.

**Comment by KvanTTT** [comment]
> Not for users (as GE does not have the GH user database), but email and the others:

But I think it's possible to bind user mention with origin link (GitHub) if this origin is single.

**Comment by mstv** [comment]
Have you noticed the links above and below the commit message in the Commit Info?
![commit info](https://user-images.githubusercontent.com/36601201/56853989-ac809c80-692f-11e9-873c-a9de5446418e.png)
The [revision links](https://git-extensions-documentation.readthedocs.io/en/latest/settings.html#git-extensions-revision-links), mentioned by Gerhard, can be extended by your own. They are configurable using regular expressions. There's also an [example GitExtensions.settings file](https://github.com/gitextensions/GitExtensionsDoc/blob/7cb9c1614b1577f66cd6c555ef2c7e52588994a9/source/files/settings/GitExtensions.settings), which you could use as a starting point.

**Comment by KvanTTT** [comment]
No, I have not noticed, thanks for pointing.

But links are recognized in the text but placed after the message, not inside.

**Comment by RussKie** [comment]
AFAIK `#1234` notation is GH specific other CVS systems don't recognise it, so rendering it as a link may be a mistake.

Rendering links as clickable can be done, however it requires additional logic what is considered a link (e.g. http|https, anything else?).

**Comment by KvanTTT** [comment]
> AFAIK #1234 notation is GH specific other CVS systems don't recognise it, so rendering it as a link may be a mistake.

At least GitLab also is able to recognize such links to issues: https://docs.gitlab.com/ee/user/project/issues/crosslinking_issues.html But it uses another syntax for merge request links: `!42`. Bitbucket also is able to: https://confluence.atlassian.com/bitbucket/mark-up-comments-305037452.html

> Rendering links as clickable can be done, however it requires additional logic what is considered a link (e.g. http|https, anything else?).

Yes, words starting with `https?://` and not escaped by \`. Also, Markdown links such as `[link title](link address)`.
